File: Blazor.BrowserAPI/BrowserAPIs/SensorAPI/Sensor/Sensor.js
/**
 * @abstract
 */
export class SensorWrapper {
    /**
     * @protected
     * @type {Sensor}
     * */
    sensor;

    /**
     * @param {Sensor} sensor
     */
    constructor(sensor) {
        this.sensor = sensor;
    }


    /**
     * @returns {boolean}
     */
    getActivated() {
        return this.sensor.activated;
    }

    /**
     * @returns {boolean}
     */
    getHasReading() {
        return this.sensor.hasReading;
    }

    /**
     * @returns {number}
     */
    getTimestamp() {
        return this.sensor.timestamp ?? 0;
    }


    /**
     */
    start() {
        this.sensor.start();
    }

    /**
     */
    stop() {
        this.sensor.stop();
    }


    // events

    /** @type {import("../../../blazor").DotNet.DotNetObject} */
    #eventTrigger;

    /** @type {boolean} */
    #isEventTriggerSync;


    // #region error event

    /**
     * @param {Event} event
     */
    #onerrorCallback = (event) => this.#isEventTriggerSync ? this.#eventTrigger.invokeMethod("InvokeError", event) : this.#eventTrigger.invokeMethodAsync("InvokeError", event);

    /**
     * @param {import("../../../blazor").DotNet.DotNetObject} eventTrigger
     * @param {boolean} isEventTriggerSync
     */
    activateOnerror(eventTrigger, isEventTriggerSync) {
        this.#eventTrigger = eventTrigger;
        this.#isEventTriggerSync = isEventTriggerSync;
        this.sensor.addEventListener("error", this.#onerrorCallback);
    }

    /**
     */
    deactivateOnerror() {
        this.sensor.removeEventListener("error", this.#onerrorCallback);
    }

    // #endregion


    // #region error event

    /**
     */
    #onactivateCallback = () => this.#isEventTriggerSync ? this.#eventTrigger.invokeMethod("InvokeActivate") : this.#eventTrigger.invokeMethodAsync("InvokeActivate");

    /**
     * @param {import("../../../blazor").DotNet.DotNetObject} eventTrigger
     * @param {boolean} isEventTriggerSync
     */
    activateOnactivate(eventTrigger, isEventTriggerSync) {
        this.#eventTrigger = eventTrigger;
        this.#isEventTriggerSync = isEventTriggerSync;
        this.sensor.addEventListener("activate", this.#onactivateCallback);
    }

    /**
     */
    deactivateOnactivate() {
        this.sensor.removeEventListener("activate", this.#onactivateCallback);
    }

    // #endregion


    // #region error event

    /**
     */
    #onreadingCallback = () => this.#isEventTriggerSync ? this.#eventTrigger.invokeMethod("InvokeReading") : this.#eventTrigger.invokeMethodAsync("InvokeReading");

    /**
     * @param {import("../../../blazor").DotNet.DotNetObject} eventTrigger
     * @param {boolean} isEventTriggerSync
     */
    activateOnreading(eventTrigger, isEventTriggerSync) {
        this.#eventTrigger = eventTrigger;
        this.#isEventTriggerSync = isEventTriggerSync;
        this.sensor.addEventListener("reading", this.#onreadingCallback);
    }

    /**
     */
    deactivateOnreading() {
        this.sensor.removeEventListener("reading", this.#onreadingCallback);
    }

    // #endregion
}

File: Blazor.BrowserAPI/BrowserAPIs/SensorAPI/Sensor/Sensor.cs
using AutoInterfaceAttributes;
using Microsoft.JSInterop;
using System.Diagnostics.CodeAnalysis;

namespace BrowserAPI.Implementation;

/// <summary>
/// The Sensor interface of the Sensor APIs is the base class for all the other sensor interfaces.
/// This interface cannot be used directly.
/// Instead it provides properties, event handlers, and methods accessed by interfaces that inherit from it.
/// </summary>
[AutoInterface(Namespace = "BrowserAPI", Inheritance = [typeof(IAsyncDisposable)])]
[RequiresUnreferencedCode("Uses Microsoft.JSInterop functionalities")]
public abstract class Sensor(IJSObjectReference sensorJS) : SensorBase(sensorJS), ISensor {
    /// <summary>
    /// Releases the JS instance for this service worker.
    /// </summary>
    public ValueTask DisposeAsync() {
        DisposeEventTrigger();
        return sensorJS.DisposeTrySync();
    }


    /// <summary>
    /// Returns a boolean value indicating whether the sensor is active.
    /// </summary>
    public ValueTask<bool> Activated => GetActivated(default);

    /// <summary>
    /// Returns a boolean value indicating whether the sensor is active.
    /// </summary>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public ValueTask<bool> GetActivated(CancellationToken cancellationToken) => sensorJS.InvokeTrySync<bool>("getActivated", cancellationToken);


    /// <summary>
    /// Returns a boolean value indicating whether the sensor has a reading.
    /// </summary>
    public ValueTask<bool> HasReading => GetHasReading(default);

    /// <summary>
    /// Returns a boolean value indicating whether the sensor has a reading.
    /// </summary>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public ValueTask<bool> GetHasReading(CancellationToken cancellationToken) => sensorJS.InvokeTrySync<bool>("getHasReading", cancellationToken);


    /// <summary>
    /// Returns the timestamp of the latest sensor reading.
    /// </summary>
    public ValueTask<double> Timestamp => GetTimestamp(default);

    /// <summary>
    /// Returns the timestamp of the latest sensor reading.
    /// </summary>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public ValueTask<double> GetTimestamp(CancellationToken cancellationToken) => sensorJS.InvokeTrySync<double>("getTimestamp", cancellationToken);



    /// <summary>
    /// Activates one of the sensors based on Sensor.
    /// </summary>
    /// <returns></returns>
    public ValueTask Start(CancellationToken cancellationToken = default) => sensorJS.InvokeVoidTrySync("start", cancellationToken);

    /// <summary>
    /// Deactivates one of the sensors based on Sensor.
    /// </summary>
    /// <returns></returns>
    public ValueTask Stop(CancellationToken cancellationToken = default) => sensorJS.InvokeVoidTrySync("stop", cancellationToken);
}
